### PR TITLE
Solr Plugins: Bibelstellensuche, Schlagwortfolgensuche

### PR DESCRIPTION
**Pull request by oobenland**

### CODE REVIEW

File: solr_plugins/Bibelstellensuche/src/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/BibleRangeFilter.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import org.apache.lucene.search.DocIdSet;
+import org.apache.lucene.search.FieldCache;
+import org.apache.lucene.search.FieldCacheDocIdSet;
+import org.apache.lucene.search.Filter;
+import org.apache.lucene.util.Bits;
+import org.apache.lucene.util.BytesRef;
+
+public class BibleRangeFilter extends Filter {
+
+	private String field = "bible_ranges";
```
Comment (ruschein) [inline solr_plugins/Bibelstellensuche/src/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/BibleRangeFilter.java]: Should there be a "final" here?

File: solr_plugins/Bibelstellensuche/src/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/BibleRangeFilter.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+		@Override
+		protected final boolean matchDoc(int doc_id) {
+			BytesRef ref = new BytesRef();
+			values.get(doc_id, ref);
+			
+			final String db_field = ref.utf8ToString();
+			if (db_field == null || db_field.isEmpty()) {
+				return false;
+			}
+			final Range[] field_ranges = BibleRangeParser
```
Comment (ruschein) [inline solr_plugins/Bibelstellensuche/src/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/BibleRangeFilter.java]: Why is this statement line-wrapped?

File: solr_plugins/Bibelstellensuche/src/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/BibleRange.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+		} else {
+			return 0;
+		}
+	}
+
+	private static int getUpper(String range) {
+		if (range.length() == 15) {
+			return Integer.valueOf(range.substring(9, 15));
+		} else {
+			return 9999999;
```
Comment (ruschein) [inline solr_plugins/Bibelstellensuche/src/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/BibleRange.java]: Don't we also have the seven nines in the string or do we really need to special-case this?

File: solr_plugins/Bibelstellensuche/src/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/BibleRangeParser.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import org.apache.solr.search.SyntaxError;
+
+public class BibleRangeParser extends QParser {
+	private final static String QUERY_SEPERATOR = " ";
+	private final static String DB_FIELD_SEPERATOR = ",";
+
+	private Query innerQuery;
+
+	public BibleRangeParser(String qstr, SolrParams localParams,
+			SolrParams params, SolrQueryRequest req) {
```
Comment (ruschein) [inline solr_plugins/Bibelstellensuche/src/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/BibleRangeParser.java]: Please indent this such that the first non-space character lines up with the first character after the opening brace above.

File: solr_plugins/Bibelstellensuche/src/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/BibleRangeParser.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+import org.apache.solr.search.QParser;
+import org.apache.solr.search.SyntaxError;
+
+public class BibleRangeParser extends QParser {
+	private final static String QUERY_SEPERATOR = " ";
+	private final static String DB_FIELD_SEPERATOR = ",";
+
+	private Query innerQuery;
+
+	public BibleRangeParser(String qstr, SolrParams localParams,
```
Comment (ruschein) [inline solr_plugins/Bibelstellensuche/src/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/BibleRangeParser.java]: qstr -> query_string

Comment (ruschein) [inline solr_plugins/Bibelstellensuche/src/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/BibleRangeParser.java]: localParams -> local_params

File: solr_plugins/Bibelstellensuche/src/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/BibleRangeParserPlugin.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+public class BibleRangeParserPlugin extends QParserPlugin {
+
+	@Override
+	@SuppressWarnings("rawtypes")
+	public void init(NamedList args) {
+	}
+
+	@Override
+	public QParser createParser(String qstr, SolrParams localParams, SolrParams params,
```
Comment (ruschein) [inline solr_plugins/Bibelstellensuche/src/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/BibleRangeParserPlugin.java]: qstr -> query_string, localParams -> local_params, req -> request

File: solr_plugins/Bibelstellensuche/src/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/BibleRangeQuery.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+import org.apache.lucene.index.AtomicReaderContext;
+import org.apache.lucene.queries.CustomScoreProvider;
+import org.apache.lucene.queries.CustomScoreQuery;
+import org.apache.lucene.search.Query;
+
+public class BibleRangeQuery extends CustomScoreQuery {
+	private final Range[] ranges;
+
+	public BibleRangeQuery(Query subQuery, Range[] ranges) {
```
Comment (ruschein) [inline solr_plugins/Bibelstellensuche/src/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/BibleRangeQuery.java]: subQuery -> sub_query

File: solr_plugins/Bibelstellensuche/src/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/BibleRangeScoreProvider.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+	public BibleRangeScoreProvider(final Range[] ranges,
+			final AtomicReaderContext context) {
+		super(context);
+		this.ranges = ranges;
+	}
+
+	@Override
+	public float customScore(final int doc_id, float subQueryScore,
+			float valSrcScores[]) throws IOException {
```
Comment (ruschein) [inline solr_plugins/Bibelstellensuche/src/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/BibleRangeScoreProvider.java]: Opening brace on next line, please.

File: solr_plugins/Bibelstellensuche/src/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/Range.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+	}
+
+	public int getMinimumDistance(Range[] targets) {
+		int query_distance = OUT_OF_RANGE;
+		for (final Range target : targets) {
+			query_distance = Math.min(getDistance(target), query_distance);
+		}
+		return query_distance;
+	}
+
```
Comment (ruschein) [inline solr_plugins/Bibelstellensuche/src/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/Range.java]: Delete empty line, please.

File: solr_plugins/Bibelstellensuche/tests/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/BibleRangeTests.java [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+		assertEquals(4, r1.getDistance(r2));
+
+		r1 = new Range(0, 2);
+		r2 = new Range(5, 7);
+		assertEquals(Range.OUT_OF_RANGE, r1.getDistance(r2));
+	}
+
+	@Test
+	public void testGetMinimumDistance() {
+		Range r1;
```
Comment (ruschein) [inline solr_plugins/Bibelstellensuche/tests/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/BibleRangeTests.java]: Combine this and the following line, please.  In general, always try to combine declaration and initialization.

File: solr_plugins/Bibelstellensuche/tests/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/BibleRangeTests.java [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
+		r1 = new BibleRange("*:*");
+		assertEquals(0, r1.getLower());
+		assertEquals(9999999, r1.getUpper());
+	}
+	
+	@Test
+	public void testDoAllSourceRangesIntersectsSomeTargetRanges() {
+		Range[] r1, r2;
+		r1 = new Range[]{new Range(1,2), new Range(5,6)};
```
Comment (ruschein) [inline solr_plugins/Bibelstellensuche/tests/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/BibleRangeTests.java]: Here, and further down, please insert a space after the comma.

File: solr_plugins/Bibelstellensuche/src/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/BibleRange.java (hunk 3 above)
Comment (oobenland) [inline solr_plugins/Bibelstellensuche/src/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/BibleRange.java]: i can't remember why, but there is a case where solr asks for *:*

File: solr_plugins/Bibelstellensuche/src/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/BibleRangeParser.java (hunk 5 above)
Comment (oobenland) [inline solr_plugins/Bibelstellensuche/src/de/uni_tuebingen/ub/ixTheo/bibleRangeSearch/BibleRangeParser.java]: I know, you don't like it, but Java is camelCase. I fixed the variable names where I did this wrong.

**Pull request merged**